cdist/exec/local.py

```diff
@@ -242,18 +246,57 @@ def run_script(self, script, env=None, return_output=False,
         return self.run(command=command, env=env, return_output=return_output,
                         message_prefix=message_prefix)
 
-    def save_cache(self):
-        destination = os.path.join(self.cache_path, self.hostdir)
+    def _cache_subpath_repl(self, matchobj):
+        if matchobj.group(2) == '%P':
+            repl = str(os.getpid())
+        elif matchobj.group(2) == '%h':
+            repl = self.hostdir
+
+        return matchobj.group(1) + repl
+
+    def _cache_subpath(self, start_time=time.time(), path_format=None):
+        if path_format:
+            repl_func = self._cache_subpath_repl
+            cache_subpath = re.sub(r'([^%]|^)(%h|%P)', repl_func, path_format)
+            dt = datetime.datetime.fromtimestamp(start_time)
+            cache_subpath = dt.strftime(cache_subpath)
+        else:
+            cache_subpath = self.hostdir
+
+        i = 0
+        while i < len(cache_subpath) and cache_subpath[i] == os.sep:
+            i += 1
+        cache_subpath = cache_subpath[i:]
+        if not cache_subpath:
+            cache_subpath = self.hostdir
+        return cache_subpath
+
+    def save_cache(self, start_time=time.time()):
+        self.log.debug("cache subpath pattern: {}".format(
+            self.cache_path_pattern))
+        cache_subpath = self._cache_subpath(start_time,
+                                            self.cache_path_pattern)
+        self.log.debug("cache subpath: {}".format(cache_subpath))
+        destination = os.path.join(self.cache_path, cache_subpath)
         self.log.debug("Saving " + self.base_path + " to " + destination)
 
-        try:
-            if os.path.exists(destination):
-                shutil.rmtree(destination)
-        except PermissionError as e:
-            raise cdist.Error(
-                    "Cannot delete old cache %s: %s" % (destination, e))
+        if not os.path.exists(destination):
+            shutil.move(self.base_path, destination)
+        else:
+            for direntry in os.listdir(self.base_path):
+                srcentry = os.path.join(self.base_path, direntry)
+                destentry = os.path.join(destination, direntry)
+                try:
+                    if os.path.isdir(destentry):
+                        shutil.rmtree(destentry)
+                    elif os.path.exists(destentry):
+                        os.remove(destentry)
+                except (PermissionError, OSError) as e:
+                    raise cdist.Error(
+                            "Cannot delete old cache entry {}: {}".format(
+                                destentry, e))
+                shutil.move(srcentry, destentry)
```

Comment:
With this change removal of destination dir entries before moving cache path (source) entries is safe.
It will not remove any directories or files not present in source path.
So if you run cdist with different cache path patterns you will not lost cache data.